### PR TITLE
Sql statement parameter

### LINKED ISSUE #30: Allow to build complex queries by bypass SqlStatement instance as parameter
Seems that it is not possible right now to do something like this:
```es6
import {SQL} from "sql-template-strings"

function prepareSet(values) { /* implementation skipped */}

pg.query(SQL`UPDATE ${SQL("table")} SET ${SQL(prepareSet({key: value})} WHERE id = ${id}`)
```

I mean that if parameter is already instance of SQLStatement, it could be inserted to query as-is (with possible merging nested parameters).

What do you think about it?

### PR DESCRIPTION
**Pull request by Toumetis**
Fixes #30 and adds the ability to perform raw substitutions in the SQL call (c.f. calling append(string))

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,14 +1,34 @@
...
    * @param {any[]} values
    */
   constructor(strings, values) {
-    this.strings = strings
-    this.values = values
+    this.strings = [strings[0]]
+    this.values = []
+    for (let i = 0; i < values.length; ++i) {
+      let value = values[i]
+      if (value instanceof SQLStatement | value instanceof RawParameter) {
```
Comment (felixfbecker) [inline index.js]: I think you want `||` here

Comment (Toumetis) [inline index.js]: Yes, thanks - fix pushed.

File: index.js [hunk 2]
```diff
@@ -83,7 +105,25 @@ function SQL(strings) {
...
+ * @param {string|string[]} strings 
+ * @param {...any} values
+ * @returns {RawParameter}
+ */
+function RAW(strings) {
+  let s  
+  if (arguments.length === 1) {
+    s = strings
+  } else {
+    s = strings.reduce((prev, curr, i) => prev + arguments[i] + curr)
```
Comment (felixfbecker) [inline index.js]: Can't you just use `String.raw.apply(String, arguments)`?

File: README.md [hunk 3]
```diff
@@ -106,6 +117,25 @@ mysql.query(SQL`SELECT * FROM `.append(mysql.escapeId(someUserInput)).append(SQL
...
+
+```js
+const RAW = SQL.RAW
+
+const table = 'books'
+const column1 = 'name'
+const column2 = 'author'
+
+const columns = RAW`${column1}, ${column2}`
+const tableName = RAW(table)
```
Comment (felixfbecker) [inline README.md]: calling a function with uppercase letters seems very unconventional to me. Template string tags seem to also move into the direction of using lowercase (there wasn't a clear convention when I started this module). I'd prefer if it was also exported as a lowercase name and we use that in the readme. We should probably add a lowercase `sql` tag too.

File: README.md [hunk 4]
```diff
@@ -106,6 +117,25 @@ mysql.query(SQL`SELECT * FROM `.append(mysql.escapeId(someUserInput)).append(SQL
 pg.query(SQL`SELECT * FROM `.append(pg.escapeIdentifier(someUserInput)).append(SQL` WHERE name = ${book} ORDER BY ${column} `).append(order))
 ```
 
+## Raw values with the RAW function / template tag
+Values returned from the RAW function (either by invocation on a string parameter or by use as a template tag) will be
+inserted raw when passed to the template literal or the `append()` function.
+
+```js
+const RAW = SQL.RAW
```
Comment (felixfbecker) [inline README.md]: I'd advocate
```js
const { raw, sql } = require('sql-template-strings')
```

**Comment by Strate** [comment]
Please grab some tests from #31 to ensure that all cases are covered. For example, I don't see test for edge case:
```
SQL`${SQL`SELECT`} * FROM table`
```

File: README.md (hunk 4 above)
Comment (rightaway) [reply README.md]: I prefer @felixfbecker's suggestion too.

File: README.md (hunk 3 above)
Comment (rightaway) [reply README.md]: @felixfbecker I agree, I was aliasing it to lowercase when using it.

File: README.md (hunk 4 above)
Comment (rightaway) [reply README.md]: I'd suggest the default export be `sql`, so that both of the following options are possible (when `import`/`export` syntax makes it into node):

```js
import sql from 'sql-template-strings'
query(sql`select ... sql.raw('something'`)
```

```js
import { sql, raw } from 'sql-template-strings'
query(sql`select ... raw('something'`)
```

**Comment by skyjur** [comment]
Development here looked stale. I haven't noticed before this, did my own implementation and opened another PR. I think I got few complex cases covered where this will break.

I appreciate work done here regarding RAW, but I think this ideally should be a separate PR.

**Comment by klandell** [comment]
Are there any status updates on this?

**Comment by Strate** [comment]
@klandell there is already prepared #31, but abandoned by @felixfbecker as "too complex"

**Comment by felixfbecker** [comment]
@Strate I understand your frustration but please understand my desire as a maintainer to keep code understandable. I didn't close your PR, I said that my gut feeling was that there must be a way to simplify. If there is none, I would be fine with merging it if the code gets some inline comments, especially the parts I pointed out as hard to understand.
